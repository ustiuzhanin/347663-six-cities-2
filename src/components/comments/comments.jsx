import React, {useEffect} from 'react';
import {connect} from 'react-redux';
import {Operations} from '../../reducer/comments/comments';
import PropTypes from 'prop-types';

/* eslint-disable camelcase*/

const Comments = (props) => {
  const {id, getComments, comments} = props;
  useEffect(() => {
    getComments(id);
  }, []);

  return (
    <>
      <h2 className='reviews__title'>
        Reviews &middot;{' '}
        <span className='reviews__amount'>{comments.length}</span>
      </h2>

      <ul className='reviews__list'>
        {comments.map((comment, i) => {
          const date = new Date(comment.date);
          const month = date.toLocaleString('default', {month: 'long'});
          const year = date.getFullYear();

          return (
            <li key={i} className='reviews__item'>
              <div className='reviews__user user'>
                <div
                  className={`reviews__avatar-wrapper user__avatar-wrapper ${comment
                    .user.is_pro && 'user__avatar-wrapper--pro'}`}
                >
                  <img
                    className='reviews__avatar user__avatar'
                    src={comment.user.avatar_url}
                    width='54'
                    height='54'
                    alt='Reviews avatar'
                  />
                </div>
                <span className='reviews__user-name'>{comment.user.name}</span>
              </div>
              <div className='reviews__info'>
                <div className='reviews__rating rating'>
                  <div className='reviews__stars rating__stars'>
                    <span
                      style={{width: `${(comment.rating / 5) * 100}%`}}
                    ></span>
                    <span className='visually-hidden'>Rating</span>
                  </div>
                </div>
                <p className='reviews__text'>{comment.comment}</p>
                <time className='reviews__time' dateTime='2019-04-24'>
                  {month} {year}
                </time>
              </div>
            </li>
          );
        })}
      </ul>
    </>
  );
};

Comments.propTypes = {
  id: PropTypes.string.isRequired,
  getComments: PropTypes.func.isRequired,
  comments: PropTypes.arrayOf(
    PropTypes.shape({
      comment: PropTypes.string.isRequired,
      date: PropTypes.string.isRequired,
      id: PropTypes.number.isRequired,
      rating: PropTypes.number.isRequired,
      user: PropTypes.shape({
        avatar_url: PropTypes.string.isRequired,
        id: PropTypes.number.isRequired,
        is_pro: PropTypes.bool.isRequired,
        name: PropTypes.string.isRequired
      }).isRequired
    })
  ).isRequired
};

export {Comments};

const mapStateToProps = (state) =>
  Object.assign({}, null, {
    comments: state.comments.comments
  });

const mapDispatchToProps = (dispatch) => ({
  getComments: (id) => {
    dispatch(Operations.loadComments(id));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(Comments);

/* eslint-enable camelcase*/
