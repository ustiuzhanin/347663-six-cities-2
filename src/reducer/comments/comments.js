const initialState = {
  comments: []
};

const ActionType = {
  LOAD_COMMENTS: 'LOAD_COMMENTS'
};

const ActionCreator = {
  loadComments: (data) => {
    return {
      type: ActionType.LOAD_COMMENTS,
      payload: data
    };
  }
};

const Operations = {
  loadComments: (id) => (dispatch, getState, api) => {
    return api.get(`/comments/${id}`).then((response) => {
      dispatch(ActionCreator.loadComments(response.data));
    });
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_COMMENTS:
      return Object.assign({}, initialState, {
        comments: action.payload
      });
  }

  return state;
};

export {ActionType, ActionCreator, Operations, reducer};
