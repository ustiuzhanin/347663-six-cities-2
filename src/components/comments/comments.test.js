import React from 'react';
import {Comments} from './comments.jsx';
import ShallowRenderer from 'react-test-renderer/shallow';

/* eslint-disable camelcase*/

test(`Comments's Snapshot`, () => {
  const renderer = new ShallowRenderer();
  renderer.render(
    <Comments
      id={'1'}
      comments={[
        {
          id: 1,
          user: {
            id: 10,
            is_pro: true,
            name: 'Max',
            avatar_url:
              'https://htmlacademy-react-2.appspot.com/six-cities/static/avatar/1.jpg'
          },
          rating: 4,
          comment:
            'The room was spacious and clean. The pool looked nothing like the photos and desparately needs a clean. The sauna and spa were closed for lunar new year holiday.',
          date: '2020-03-19T00:32:20.983Z'
        }
      ]}
      getComments={jest.fn()}
    />
  );
  const result = renderer.getRenderOutput();

  expect(result).toMatchSnapshot();
});
/* eslint-enable camelcase*/
