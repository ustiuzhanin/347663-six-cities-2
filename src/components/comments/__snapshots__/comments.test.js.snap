// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments's Snapshot 1`] = `
<React.Fragment>
  <h2
    className="reviews__title"
  >
    Reviews Â·
     
    <span
      className="reviews__amount"
    >
      1
    </span>
  </h2>
  <ul
    className="reviews__list"
  >
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper user__avatar-wrapper--pro"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="https://htmlacademy-react-2.appspot.com/six-cities/static/avatar/1.jpg"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Max
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "80%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          The room was spacious and clean. The pool looked nothing like the photos and desparately needs a clean. The sauna and spa were closed for lunar new year holiday.
        </p>
        <time
          className="reviews__time"
          dateTime="2019-04-24"
        >
          March
           
          2020
        </time>
      </div>
    </li>
  </ul>
</React.Fragment>
`;
